test/test-tcp-close-reset.c

```diff
@@ -0,0 +1,293 @@
+/* Copyright libuv project contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <errno.h>
+#include <string.h> /* memset */
+
+#define NUM_WRITE_REQS 4
+
+static uv_loop_t* loop;
+static uv_tcp_t tcp_server;
+static uv_tcp_t tcp_client;
+static uv_tcp_t tcp_accepted;
+static uv_connect_t connect_req;
+static uv_shutdown_t shutdown_req;
+
+static int client_close;
+static int shutdown_before_close;
+
+static int write_cb_called;
+static int close_cb_called;
+static int shutdown_cb_called;
+
+static void connect_cb(uv_connect_t* req, int status);
+static void write_cb(uv_write_t* req, int status);
+static void close_cb(uv_handle_t* handle);
+static void shutdown_cb(uv_shutdown_t* req, int status);
+
+static int read_size;
+
+
+static void do_write(uv_tcp_t* handle) {
+  uv_write_t* req;
+  uv_buf_t buf;
+  int i, r;
+
+  buf = uv_buf_init("PING", 4);
+  for (i = 0; i < NUM_WRITE_REQS; i++) {
+    req = malloc(sizeof *req);
```

Comment:
I'd use a `static uv_write_t write_reqs[4];` for this and then check `i < ARRAY_SIZE(write_reqs)`, or do:
```c
for (req = write_reqs; req < ARRAY_END(write_reqs); req++) {
  /* ... */
}
```

If we don't have an `ARRAY_END()` macro already, its definition is this:
```c
#define ARRAY_END(x) ((x) + ARRAY_SIZE(x))
```

Reply:
Yes, I was lazy and copied part of the code from other tests. Will improve it per your suggestions.